### PR TITLE
RFC 1008: TLS and ALPN

### PR DESCRIPTION
**Pull request by fantix**
* Load certs from `data_dir` if no `--tls-cert-file` given
* Generate dev certs if none found in dev mode
* ~Run tests using `_localtest_PORT` temporary instances~

Closes #2610

### CODE REVIEW

File: edb/server/args.py [hunk 1]
```diff
@@ -316,6 +320,23 @@ def _writer(status: str) -> None:
...
+             'self-signed certificate will be generated and used instead (for '
+             'development only!).'),
+    click.option(
+        '--tls-key-file', type=PathPath(),
+        help='Specify a path to a file containing the private key. If not '
+             'present, the private key will be taken from --tls-cert-file as '
+             'well. If the private key is protected by a password, specify '
+             'with an environment variable EDGEDB_TLS_PRIVATE_KEY_PASSWORD.'),
+    click.option(
+        '--optional-tls', type=bool, is_flag=True, hidden=True,
```
Comment (elprans) [inline edb/server/args.py:337]: I'd name this `--allow-cleartext-connections` for clarity.

File: edb/server/cluster.py [hunk 2]
```diff
@@ -245,6 +248,7 @@ async def test(timeout):
                     conn = await edgedb.async_connect(
                         host=str(self._runstate_dir),
                         port=self._effective_port,
+                        admin=True,
```
Comment (elprans) [inline edb/server/cluster.py:251]: We are deprecating and removing the `admin` argument in bindings.  Why is this necessary?

File: edb/server/main.py [hunk 3]
```diff
@@ -217,6 +223,50 @@ async def _run_server(
...
+                    "_generate_dev_cert",
+                    "--cert-file",
+                    str(cert_file),
+                    "--key-file",
+                    str(key_file),
+                ]
+            )
+            return str(cert_file), str(key_file)
+
+    abort("Cannot start EdgeDB server without TLS certificate.")
```
Comment (elprans) [inline edb/server/main.py:267]:
```suggestion
    abort("Cannot start EdgeDB server without a TLS certificate, use the `--tls-cert-file` command-line argument to specify it.")
```

File: edb/server/server.py [hunk 4]
```diff
@@ -811,36 +819,33 @@ async def run_startup_script_and_exit(self):
     async def _start_servers(self, host, port):
         nethost = await _fix_localhost(host)
 
+        sslctx = ssl.SSLContext()
+        sslctx.load_cert_chain(
```
Comment (elprans) [inline edb/server/server.py:823]: We should handle cert issues here and wrap them with a nice error.  Specifically, if `tls_key_file` is missing and not present in `cert_file`.

File: edb/server/server.py [hunk 5]
```diff
@@ -1010,3 +1016,7 @@ async def _fix_localhost(host):
         hosts.insert(idx, addr)
 
     return hosts
+
+
+def _tls_private_key_password():
+    return os.environ.get('EDGEDB_TLS_PRIVATE_KEY_PASSWORD', '')
```
Comment (elprans) [inline edb/server/server.py:1022]: Let's use `EDGEDB_SERVER` as the var prefix.

File: edb/server/cluster.py (hunk 2 above)
Comment (fantix) [reply edb/server/cluster.py:251]: It's not - it's just easier to make the tests pass for now, given that we still have a few `_admin_query()` calls. I can fix those to use the CLI instead of the Python binding.

File: edb/server/server.py [hunk 6]
```diff
@@ -852,13 +856,58 @@ async def _start_servers(self, host, port):
...
+                        "the private key file using the `--tls-key-file` "
+                        "command-line argument?"
+                    ) from e
+                else:
+                    raise StartupError(
+                        "Cannot load TLS certificates - please double check "
+                        "if the specified certificate files are valid."
+                    )
+
+            raise StartupError("") from e
```
Comment (elprans) [inline edb/server/server.py:906]:
```suggestion
            raise StartupError from e
```

File: edb/server/protocol/protocol.pyx [hunk 7]
```diff
@@ -189,13 +231,89 @@ cdef class HttpProtocol:
...
+        else:
+            # It's either HTTP as the negotiated protocol, or the negotiation
+            # failed and we have no idea what ALPN the client has set. Here we
+            # just start talking in HTTP, and let the client bindings decide if
+            # this is an error based on the ALPN result.
+            self._init_http_parser()
+
+    async def _return_binary_error(self, proto):
+        await proto.write_error(errors.BinaryProtocolError(
+            'TLS Required',
```
Comment (elprans) [inline edb/server/protocol/protocol.pyx:283]: Can we add a test for this condition?  `edgedb-python` auto-downgrades to cleartext on cert errors, we can utilize that.

**Review by elprans — APPROVED** [review]
LGTM overall.  @1st1, please take a look too.

File: edb/server/args.py [hunk 8]
```diff
@@ -316,6 +320,22 @@ def _writer(status: str) -> None:
         '--auto-shutdown-after', type=float, default=-1.0,
         help='shutdown the server after the last connection has been closed '
              'for N seconds. N < 0 is treated as infinite.'),
+    click.option(
+        '--tls-cert-file', type=PathPath(),
+        help='Specify a path to a single file in PEM format containing the TLS'
```
Comment (1st1) [inline edb/server/args.py:325]:
```suggestion
        help='Specify a path to a single file in PEM format containing the TLS '
```

File: edb/server/args.py [hunk 9]
```diff
@@ -316,6 +320,22 @@ def _writer(status: str) -> None:
...
+             'certificate as well as any number of CA certificates needed to '
+             'establish the certificate’s authenticity. If not present, a '
```
Comment (1st1) [inline edb/server/args.py:327]: I'd do this differently.

`edb server` should have `--tls-cert-file` optional.
`edgedb-server` should have it required.

With adjusted help section for each tool.

File: edb/server/args.py [hunk 10]
```diff
@@ -316,6 +320,22 @@ def _writer(status: str) -> None:
...
+             'self-signed certificate will be generated and used instead (for '
+             'development only!).'),
+    click.option(
+        '--tls-key-file', type=PathPath(),
+        help='Specify a path to a file containing the private key. If not '
+             'present, the private key will be taken from --tls-cert-file as '
+             'well. If the private key is protected by a password, specify '
+             'with an environment variable EDGEDB_TLS_PRIVATE_KEY_PASSWORD.'),
```
Comment (1st1) [inline edb/server/args.py:335]:
```suggestion
             'with an environment variable EDGEDB_SERVER_TLS_PRIVATE_KEY_PASSWORD.'),
```

File: edb/server/cluster.py [hunk 11]
```diff
@@ -216,69 +218,59 @@ async def _read_server_status(stream: asyncio.StreamReader):
...
+            raise ClusterError(
+                f'could not connect to edgedb-server '
+                f'within {timeout} seconds') from None
+
+    def _admin_query(self, query, wait_until_available="0s"):
+        return subprocess.call(
+            [
+                sys.executable,
+                "-m",
+                "edb.cli",
```
Comment (1st1) [inline edb/server/cluster.py:256]: I think we should call the `edgedb` command here explicitly; `edb.cli` is just a wrapper.

File: edb/server/main.py [hunk 12]
```diff
@@ -217,6 +223,51 @@ async def _run_server(
...
 
+def _init_tls_certs(
+    args: srvargs.ServerConfig,
+    cert_dir: Optional[tempfile.TemporaryDirectory],
+):
+    cert_path = None
+    if args.tls_cert_file:
+        cert_file = pathlib.Path(args.tls_cert_file).resolve()
+        if args.tls_key_file is None:
+            return str(cert_file), None
```
Comment (1st1) [inline edb/server/main.py:234]: I'm just curious -- why not just return Path objects?

File: edb/server/protocol/protocol.pyx [hunk 13]
```diff
@@ -189,13 +231,89 @@ cdef class HttpProtocol:
...
+            binproto.data_received(data)
+        return binproto
+
+    def _init_http_parser(self):
+        self.parser = httptools.HttpRequestParser(self)
+        self.current_request = HttpRequest()
+
+    async def _forward_first_data(self, data):
+        # As we stole the "first data", we need to manually send it back to
+        # the SSLProtocol. The hack here is highly-coupled with uvloop impl.
```
Comment (1st1) [inline edb/server/protocol/protocol.pyx:248]: Perhaps we should add an explicit method for this in uvloop, like `inject_data` or something. But that can be done separately if you think it's a good idea.

**Review by 1st1 — APPROVED** [review]
Great work! I've left some comments.

File: edb/server/main.py (hunk 12 above)
Comment (fantix) [reply edb/server/main.py:234]: Oh yes, `Path` objects would do better! It's been a few versions of changes and I cannot remember how it ended up in strings. Let me push a fix.

File: edb/server/cluster.py (hunk 11 above)
Comment (fantix) [reply edb/server/cluster.py:256]: But the `edgedb` command is not always in the `PATH` - finding the path to the binary is the same code as `edb.cli`(`.__main__`) I think?

Comment (elprans) [reply edb/server/cluster.py:256]: This method is only used in tests to configure test clusters, using `edb.cli` here is appropriate.

File: edb/server/protocol/protocol.pyx (hunk 7 above)
Comment (fantix) [reply edb/server/protocol/protocol.pyx:283]: > `edgedb-python` auto-downgrades to cleartext on cert errors

Oh it's actually the opposite - it auto-downgrades to cleartext if the `SSLError` is not a cert error. That means the server is likely an old one without TLS support, and the client binding is allowed to use cleartext.

I just added 2 tests using `edb.protocol.protocol`, that should cover the case.

File: edb/server/server.py (hunk 6 above)
Comment (fantix) [reply edb/server/server.py]: Oops my bad - missed one error message here.

File: edb/server/args.py (hunk 9 above)
Comment (fantix) [reply edb/server/args.py]: Had to add a hidden `edb testserver` sub-command for the tests to leverage the cert auto-generation feature, at least the tests passed now.

Comment (elprans) [reply edb/server/args.py]: What is the reason for avoiding generating dev certs in `edb server`?

Comment (fantix) [reply edb/server/args.py]: Oh we also generate dev certs in `edb server`, it's just that neither `edb server` nor `edgedb-server` (`python -m edgedb.server.main`) is suitable for running the tests. `edb server` has `insecure=True` and `edgedb-server` doesn't generate the certs.

File: edb/server/cluster.py (hunk 11 above)
Comment (fantix) [reply edb/server/cluster.py:256]: Ohhh this is biting the nightly build test where we don't have the `edgedb` CLI binary. I'll change it to use `edb.protocol.protocol` to connect to the UNIX socket directly and run admin queries there.

Comment (elprans) [reply edb/server/cluster.py:256]: We actually have it in `PATH` there, so it might be easier to simply do use unprefixed "edgedb" here?

Comment (fantix) [reply edb/server/cluster.py:256]: Ah yes, sure thing!
